core/src/main/java/org/infinispan/stream/impl/ClusterStreamManagerImpl.java

```diff
@@ -0,0 +1,438 @@
+                  boolean wasSuspect = false;
+                  // Unwrap exception
+                  do {
+                     if (cause instanceof SuspectException) {
+                        log.tracef("Exception from %s contained a SuspectException, making all segments %s suspect",
+                                dest, targetSegments);
+                        receiveResponse(uuid, dest, true, targetSegments, null);
+                        wasSuspect = true;
+                        break;
+                     }
+                  } while ((cause = cause.getCause()) != null);
+
+                  if (!wasSuspect) {
+                     log.tracef(e, "Encounted exception for %s from %s", uuid, dest);
+                     throw e;
+                  }
+               }
+            }
+         }
+      }
+      return uuid;
+   }
+
+   private void submitAsyncTasks(UUID uuid, Map<Address, Set<Integer>> targets, Map<Integer, Set<K>> keysToExclude,
+                                 boolean parallelStream, Set<K> keysToInclude, boolean includeLoader,
+                                 StreamRequestCommand.Type type, Object operation) {
+      for (Map.Entry<Address, Set<Integer>> targetInfo : targets.entrySet()) {
+         Set<Integer> segments = targetInfo.getValue();
+         Set<K> keysExcluded = determineExcludedKeys(keysToExclude, segments);
+         Address dest = targetInfo.getKey();
+         log.tracef("Submitting async task to %s for %s excluding keys %s", dest, uuid, keysExcluded);
+         CompletableFuture<Map<Address, Response>> completableFuture = rpc.invokeRemotelyAsync(
+                 Collections.singleton(dest), factory.buildStreamRequestCommand(uuid, parallelStream, type, segments,
+                         keysToInclude, keysExcluded, includeLoader, operation),
+                 rpc.getDefaultRpcOptions(true));
+         completableFuture.whenComplete((v, e) -> {
+            if (v != null) {
+               Response response = v.values().iterator().next();
+               if (!response.isSuccessful()) {
+                  log.tracef("Unsuccessful response for %s from %s - making segments suspect", uuid, targetInfo.getKey());
+                  receiveResponse(uuid, targetInfo.getKey(), true, targetInfo.getValue(), null);
+               }
+            } else if (e != null) {
+               Throwable cause = e;
+               boolean wasSuspect = false;
+               // Unwrap the exception
+               do {
+                  if (cause instanceof SuspectException) {
+                     log.tracef("Exception contained a SuspectException, making all segments %s suspect",
+                             targetInfo.getValue());
+                     receiveResponse(uuid, targetInfo.getKey(), true, targetInfo.getValue(), null);
+                     wasSuspect = true;
+                     break;
+                  }
+               } while ((cause = cause.getCause()) != null);
```

Comment:
I'm not a big fan of these iterations searching for a cause, because it makes me feel that we don't know exactly what's going on and we need to leave ourselves some wiggle room. The CompletionExceptions do muddy the situation a bit, so I'm ok if you don't want to change it, but I'd suggest at least extracting the search for the SuspectException in a separate method.